### PR TITLE
Document abtract models

### LINKED ISSUE #85: Abstract / Swappable models
Hi,

are you open to the idea of making the logic in the models of django-plans available as abstract / swappable models (using [django-swappable](https://github.com/wq/django-swappable-models))?

We've been using this approach successfully in some modules of the [OpenWISP project](http://openwisp.org/), eg: [django-freeradius](https://github.com/openwisp/django-freeradius).

OpenWISP needs a way to provide features like premium service and periodic paid subscriptions with expiration. I've been reading the code of this package and another package named silver but I'm more inclined to use this package because it feels a lot simpler. The problem is that obviously I will need to customize a few things to integrate it nicely into OpenWISP and the best way to do that without hacks is to have abstract models.

I did something similar a while ago for django-organizations: https://github.com/bennylope/django-organizations/pull/119.

If you are open to the idea I'm proposing I'm available to join forces with you and improve this package.

### PR DESCRIPTION
**Pull request by NoumbissiValere**
Closes #85

### CODE REVIEW

File: docs/source/settings.rst
Comment (nemesisdesign) [inline docs/source/settings.rst:20]: we should specify that `custom_plans` is a django app, something like:

Given a django app named `custom_plans` which provides custom models for django-plans, these are the settings to use:

...

File: setup.py
Comment (nemesisdesign) [inline setup.py:39]: please bind the version

File: docs/source/customize_models.rst
Comment (nemesisdesign) [inline docs/source/customize_models.rst:32]: fix the variable setting of all the vars please

File: docs/source/customize_models.rst
Comment (nemesisdesign) [inline docs/source/customize_models.rst:1]: let's call this `Extending django-plans: custom models`

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

**Comment by Alir3z4** [comment]
@NoumbissiValere thanks for the great work.
Could you please add tests to cover all the new additions, including their expected behavior and some edge cases?

**Comment by nemesisdesign** [comment]
@Alir3z4 thanks for your feedback, here's what we can do (which is what we do in OpenWISP):

- we can add a test app with the extended models (eg: https://github.com/openwisp/openwisp-radius/tree/master/tests/openwisp2/sample_radius)
- we can edit the existing codebase to load models using `swappable.load_model()`, this should allow us to run the same tests also with the extended models
- we can add some logic in the `settings.py` which runs the customized models when an env variable like `SAMPLE_APP=1` is present
- then we can run again tests for the sample app

The expected behavior is that with custom models which do not alter the core behavior of django-plans, the app keeps working fine.

Regarding additional test cases, there's no really much to do apart from maybe we could add an additional field and add a test which checks for the presence of this additional field in the test model.

For edge cases: what do you mean exactly? This is a bit broad.
We focus on what we need, which is a way to extend the behavior of base apps without having to fork the core code.
We've been doing similar improvements to other libraries as well, see:

- https://github.com/django-notifications/django-notifications/pull/255
- https://github.com/django-notifications/django-notifications/pull/284
- https://github.com/bennylope/django-organizations/pull/119

Let us know if this would be enough for getting the PR merged.

**Comment by Alir3z4** [comment]
Sorry for the long delay to get back to this.

Regarding the testing the application, I don't have any specific preferences here. As far as I know, dealing with Abstract models would be better to lead to a test app.

As long as the new additions have a test to show case the functionality and a proof that things work as expected this should be good to go.

It's important to keep a healthy code coverage here. Unfortunately Travis-ci is shutting down for open source projects and may not longer our tests, therefore there won't be any code coverage here. Feel free to submit a patch for switching to Github Actions to get the tests running again, either in this pull request or a separate one (_probably I'll do it, but I'm not sure when I'll get a good time to do it_).

**Comment by nemesisdesign** [comment]
> Sorry for the long delay to get back to this.
> 
> Regarding the testing the application, I don't have any specific preferences here. As far as I know, dealing with Abstract models would be better to lead to a test app.
> 
> As long as the new additions have a test to show case the functionality and a proof that things work as expected this should be good to go.
> 
> It's important to keep a healthy code coverage here. Unfortunately Travis-ci is shutting down for open source projects and may not longer our tests, therefore there won't be any code coverage here. Feel free to submit a patch for switching to Github Actions to get the tests running again, either in this pull request or a separate one (_probably I'll do it, but I'm not sure when I'll get a good time to do it_).

We'll ensure the code coverage does not decrease but we won't be able to help with github actions right now. We have our own project @openwisp we maintain where we have the same problem with multiple repositories and we are busy fixing those ones.

**Comment by pandafy** [comment]
Hey @Alir3z4! I have rebased this PR on the current master. I have locally compared coverage and it appeared to be consisted with the current master. Will you please approve the workflow run?

**Comment by pandafy** [comment]
@Alir3z4 I have added a sample app in demo/example (sample_plans) to test extensibility of the models.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
@pandafy this needs to be rebased on the latest master.

**Comment by Alir3z4** [comment]
@PetrDlouhy can you do a quick review on this please if you can?

**Comment by PetrDlouhy** [comment]
I looked at this briefly. The overall direction looks very good, although I was not able to see all the changes in such a long diff.
The tests are not passing, and the code should be rebased to the current master.
There are also commits like [Django 4 compatibility](https://github.com/django-getpaid/django-plans/pull/126/commits/6bd8c38dc766d7e6e69d2ad1ed61bae4b6401735) or [Removed python_2_unicode_compatible decorator from models](https://github.com/django-getpaid/django-plans/pull/126/commits/a7959f5592534f2f4c6bf092338ed1d08b0bf448) which are probably already solved in current master (or should be squashed with the previous commits).

I would also prefer, if I could see diff between `plans/models.py` and `plans/base/models.py`. Most of the code moved there, so it would make sense that it done with `git mv`. That would probably also make rebasing easyer.

**Comment by pandafy** [comment]
Thanks for your review @PetrDlouhy 😄 

I have rebased this PR on the current master branch. 

> There are also commits like [Django 4 compatibility](https://github.com/django-getpaid/django-plans/pull/126/commits/6bd8c38dc766d7e6e69d2ad1ed61bae4b6401735) or [Removed python_2_unicode_compatible decorator from models](https://github.com/django-getpaid/django-plans/pull/126/commits/a7959f5592534f2f4c6bf092338ed1d08b0bf448) which are probably already solved in current master (or should be squashed with the previous commits).

I think we can squash all of these commits into one at the end. I created separate commits to aid reviews. Let me know, if I should squash them now. 

> I would also prefer, if I could see diff between `plans/models.py` and `plans/base/models.py`. Most of the code moved there, so it would make sense that it done with `git mv`. That would probably also make rebasing easyer.

I tried using `git mv` but it created more merge conflicts. You can check the diffs on this URL: https://www.diffchecker.com/MOeYFIPC

**Comment by PetrDlouhy** [comment]
@pandafy I see few lines, that were reintroduced in your diff of models.py:
* Line `tax=userplan.recurring.tax,` in `RecurringUserPlan.create_renew_order` was removed in https://github.com/django-getpaid/django-plans/commit/80f0c667e526b9520cc31d91ed00eb77f0419439
* You are removing `get_plan_pricing()` method

**Comment by nemesisdesign** [comment]
> @PetrDlouhy thank you for your another round of review. I have made those changes. Here is the updated diff: https://www.diffchecker.com/bYdFQ9Rq

Thank you @pandafy! @PetrDlouhy let us know if you think anything else is missing.

**Comment by nemesisdesign** [comment]
> > @PetrDlouhy thank you for your another round of review. I have made those changes. Here is the updated diff: https://www.diffchecker.com/bYdFQ9Rq
> 
> Thank you @pandafy! @PetrDlouhy let us know if you think anything else is missing.

Ping @Alir3z4, please let us know if you think this can be merged or not, we invested quite a good amount of hours on it and it's not yet clear if it's good enough, if it needs to be improved, if it will be merged or not.

I want to point out it's been 4 years I have been trying to merge this PR:
https://github.com/django-getpaid/django-plans/issues/85

Please let us know.

**Comment by PetrDlouhy** [comment]
@nemesisdesign @Alir3z4 I have added tests to all views in #165, so it might be a good idea to merge that one first along with #163.
The rebase should be quite forward and the extended tests would show if there are any more potential problems.

(Don't merge #161 and #162 before this. The merge could be painful and I can rebase afterwards.)

**Comment by nemesisdesign** [comment]
@Alir3z4 is there any chance we can advance with this? Do you need a helping hand in maintaining this package?

**Comment by Alir3z4** [comment]
> I have added tests to all views in https://github.com/django-getpaid/django-plans/pull/165, so it might be a good idea to merge that one first along with https://github.com/django-getpaid/django-plans/pull/163.

Merged them just now.

@nemesisdesign shouldn't the open discussion on the code be resolved first before merging?
If they're not that important, we can go ahead and merge,

Once marked as resolved, I'll merge right away.

File: docs/source/settings.rst
Comment (pandafy) [reply docs/source/settings.rst:21]: This is resolved

**Comment by nemesisdesign** [comment]
@Alir3z4 we're ready, all review comments have been resolved, we cannot flag those as resolved, you can merge now :+1: :+1: 
Thanks @pandafy!

**Comment by Alir3z4** [comment]
Sorry, I've missed the latest comment about being ready.
It's in.

Excellent contribution has been done here!
Thanks to everyone that made this happen <3

**Comment by anuj9196** [comment]
I'm so glad the plugin has this feature but I'm just worried about its future. The dependency used to implement this change was last updated on *Released: Nov 30, 2021*. 

Would you happen to have any plan if the dependent library becomes obsolete and not suitable for future versions of Django?

**Comment by pandafy** [comment]
@anuj9196 I am one of the maintainers of swapper and we use it extensively at @openwisp. 

Swappable models have been supported in Django for sometime but as a private API. You can read [this issue on Django's Trac](https://code.djangoproject.com/ticket/19103) for more details.

**Pull request merged**
